thirdparty/hnswlib/hnswlib/neighbor.h

```diff
@@ -28,83 +28,116 @@ struct Neighbor {
     }
 };
 typedef std::priority_queue<Neighbor, std::vector<Neighbor>, std::greater<Neighbor>> IteratorMinHeap;
-class NeighborSet {
- public:
-    explicit NeighborSet(size_t capacity = 0) : capacity_(capacity), data_(capacity_ + 1) {
-    }
 
-    // will push any neighbor that does not fit into NeighborSet to disqualified.
-    // When searching for iterator, those points removed from NeighborSet may be
-    // qualified candidates as the iterator iterates, thus we need to retain
-    // instead of disposing them.
-    bool
-    insert(Neighbor nbr, IteratorMinHeap* disqualified = nullptr) {
-        if (size_ == capacity_ && nbr.distance >= data_[size_ - 1].distance) {
-            if (disqualified) {
-                disqualified->push(nbr);
-            }
-            return false;
+template <bool need_save>
+class NeighborSetPopList {
+ private:
+    inline auto
+    find_binsert_position(size_t lo, size_t hi, float distance) const -> size_t {
+        if (lo >= hi) {
+            return lo;
         }
-        if (size_ == capacity_ && disqualified) {
-            disqualified->push(data_[size_ - 1]);
+        if (data_[lo].distance >= distance) {
+            return lo;
+        }
+        if (data_[hi - 1].distance <= distance) {
+            return hi;
         }
-        int lo = 0, hi = size_;
         while (lo < hi) {
```

Comment:
can this loop be replaced with either `std::upper_bound()` or `std::lower_bound()`?

Reply:
Good idea~ I will use `upper_bound`.